File: Src/java/engine-jackson/src/test/java/org/opencds/cqf/cql/engine/execution/CqlExternalFunctionsTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.util.Arrays;

import org.opencds.cqf.cql.engine.data.SystemExternalFunctionProvider;
import org.opencds.cqf.cql.engine.execution.external.MyMath;
import org.testng.annotations.Test;

public class CqlExternalFunctionsTest extends CqlExecutionTestBase {

    @Test
    public void testExternalFunctions() {
        Context context = new Context(library);

        context.registerExternalFunctionProvider(
            library.getIdentifier(),
            new SystemExternalFunctionProvider(Arrays.asList(MyMath.class.getDeclaredMethods()))
        );

        Object result;

        result = context.resolveExpressionRef("CallMyPlus").getExpression().evaluate(context);
        assertThat(result, is(10));

        result = context.resolveExpressionRef("CallMyMinus").getExpression().evaluate(context);
        assertThat(result, is(-2));
    }
}

File: Src/java/engine/src/jvmTest/kotlin/org/opencds/cqf/cql/engine/execution/CqlExternalFunctionsTest.kt
package org.opencds.cqf.cql.engine.execution

import kotlin.test.Test
import kotlin.test.assertEquals
import org.opencds.cqf.cql.engine.data.StaticFunction
import org.opencds.cqf.cql.engine.data.SystemExternalFunctionProvider
import org.opencds.cqf.cql.engine.execution.external.MyMath
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlInteger

internal class CqlExternalFunctionsTest : CqlTestBase() {
    @Test
    fun externalFunctions() {
        val identifier = toElmIdentifier("CqlExternalFunctionsTest")

        val engine = engine
        engine.state.environment.registerExternalFunctionProvider(
            identifier,
            SystemExternalFunctionProvider(
                MyMath::class.java.declaredMethods.map {
                    StaticFunction(
                        it.name,
                        { arguments ->
                            it.invoke(it.declaringClass, *arguments!!.toTypedArray()) as Value?
                        },
                    )
                }
            ),
        )

        val results = engine.evaluate { library(identifier) }.onlyResultOrThrow
        var value = results["CallMyPlus"]!!.value
        assertEquals(10.toCqlInteger(), value)

        value = results["CallMyMinus"]!!.value
        assertEquals((-2).toCqlInteger(), value)
    }
}
